### ISSUE TITLE
Question about updated dependencies

### ISSUE THREAD

**Issue by rambius** [conan-io/conan#4997]
Hello,

I have a question about handling dependencies with conan. Let's say I have a repository in an artifactory instance (or conan-server running somehre). My project P that has a dependency D with version V, specified in conanfile.txt or conanfile.py. conan install fetches that version V to the local cache on the local machine.

Now suppose I have discovered a bug in D's recipe. I fix it and run conan create followed by conan upload to get the fix in the repository. This does not upload a new version of D, it is the same version. Will conan install sense the change and re-download the dependency even if it is in the cache with the same version? Or do I have to wipe the local cache to force the download?

Regard
rambius

**Comment by Johnnyxy** [comment]
Hi there,
if the recipe differs from the one on the server where it is uploaded to, conan will detect that it differs and will upload the recipe. 
This will purge the recipe and all packages "owned" by the recipe from the remote repository and replace them with the new recipe. Additionally conan can upload built packages if one specifies ```--all``` ([see](https://docs.conan.io/en/latest/reference/commands/creator/upload.html#conan-upload)).

Any clients who use ```--update``` with ```conan install``` will automatically fetch the new recipe and packages if an appropriate package is available.

Check the output of ```conan install``` on the client side of conan:
```
Requirements
    ...
Packages
    ...
Build requirements
    ...
Build requirements packages
    ...
```


You cannot replace the recipe without purging the packages in the any conan repository. This ensures that all packages in the conan repository are synchronized with the recipe.
There is an exception to this but this is only for development purposes ([see here](https://docs.conan.io/en/latest/reference/commands/creator/export.html#conan-export)).

**Comment by rambius** [comment]
Thank you for your explanation.

**Comment by memsharded** [comment]
Thanks @Johnnyxy!

I would like to contribute with some further details about the new "revisions" feature that has been introduced in last Artifactory release (together with Conan 1.14):

- With revisions "enabled" (opt-in), when you do a change in a recipe, it doesn't overwrite the previous recipe, but it creates an internal automatic version of the recipe, called the "revision" of the recipe, by default, it is the hash of the recipe contents. 
- With revisions enabled, the server stores all the revisions, they don't overwrite when uploaded
- Revisions resolve automatically and transparently to the latest revision, but it is possible to also explicitly select a specific revision, like ``pkg/version@user/channel:revision``.
- They are also updated with the ``conan install --update`` argument

In short: "revisions" is an interesting feature if you want to do changes to recipe/packages without changing/bumping its version, but also achieve better immutability and reproducibility.  More about it here: https://docs.conan.io/en/latest/mastering/revisions.html
